### PR TITLE
Fix cluster version check for PSP and use it in pspdelete controller registration for RKE2

### PR DESCRIPTION
**Pull request by maxsokolovsky**
## Problem

When admins upgrade a downstream RKE2 cluster that has the PSP feature (version <1.25) to one that doesn't (version >=1.25) and still has a non-empty value for `defaultPodSecurityPolicyTemplateName` in cluster configuration, there are two issues.
1. Recurring controller failures to `list *v1beta1.PodSecurityPolicy: the server could not find the requested resource`.
2. Many handlers do not run - Rancher isn't creating LimitRanges in project namespaces that have container default limits, isn't creating v3.Tokens when users download their Kubeconfigs, and so on.
 
## Solution

1. Fix the cluster version check for compatibility of remote clusters with PSPs. Make it return an error for clusters v1.25+ and whose version string has additional characters, like `v1.26.9+rke2r1`.
2. Ensure the `pspdelete` controller also doesn't get registered if Rancher attempts to do so for a cluster whose version doesn't pass the cluster version check. Rancher tries to register this controller only for RKE2 clusters.
 
### Automated Testing

* Test types added/modified:
    * Unit (for the cluster version check)
    * Built a custom Rancher image and tried reproducing the main issues 

### Regressions Considerations

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

### CODE REVIEW

**Comment by samjustus** [comment]
#43202

**Comment by gregsidelinger** [comment]
So this looks to address the two issues around PSPs when they don't exist in 1.25+ but any idea why this is breaking the controller reconcile loops for the other objects?

**Comment by maxsokolovsky** [comment]
This is the main mystery. I will consult with someone who is an expert on controllers and caches.

File: pkg/controllers/managementuser/rbac/podsecuritypolicy/serviceaccount.go [hunk 1]
```diff
@@ -118,12 +118,12 @@ func (m *serviceAccountManager) sync(key string, obj *v1.ServiceAccount) (runtim
...
 
-	err := checkClusterVersion(m.clusterName, m.clusterLister)
+	err := CheckClusterVersion(m.clusterName, m.clusterLister)
 	if err != nil {
-		if errors.Is(err, errVersionIncompatible) {
+		if errors.Is(err, ErrClusterVersionIncompatible) {
 			return obj, nil
 		}
-		return obj, fmt.Errorf(clusterVersionCheckErrorString, err)
+		return obj, err
```
Comment (MbolotSuse) [inline pkg/controllers/managementuser/rbac/podsecuritypolicy/serviceaccount.go:126]: If you are going to make changes to how this error is handled, I would recommend that you properly wrap the error, something like:

```suggestion
		return obj, fmt.Errorf("error when checking cluster version for serviceAccount controller: %w", err)
```

File: pkg/controllers/managementuser/rbac/podsecuritypolicy/version.go [hunk 2]
```diff
@@ -4,25 +4,33 @@ import (
...
-var clusterVersionCheckErrorString = "failed to check cluster version compatibility for podsecuritypolicy controllers: %v"
+var ErrClusterVersionIncompatible = errors.New("podsecuritypolicies are not available in Kubernetes v1.25 and above")
 
-// checkClusterVersion tries to fetch a cluster by name, extract its Kubernetes version,
-// and check if the version is less than v1.25.
-func checkClusterVersion(clusterName string, clusterLister v3.ClusterLister) error {
+// CheckClusterVersion tries to fetch a cluster by name, extract its Kubernetes version,
+// and check if the cluster supports PodSecurityPolicies.
+func CheckClusterVersion(clusterName string, clusterLister v3.ClusterLister) error {
+	const mainErr = "failed to check cluster version compatibility for podsecuritypolicy controllers"
```
Comment (MbolotSuse) [inline pkg/controllers/managementuser/rbac/podsecuritypolicy/version.go:16]: I don't think you are getting much out of this variable, I would recommend forming specific error messages below and not relying on a "primary" error message.

File: pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+			version: "v1.25.9-rancher1-1",
+			wantErr: true,
+		},
+		{
+			version: "v1.26.9-rancher1-1",
+			wantErr: true,
+		},
+		// rke2 version strings
+		{
+			version: "v1.24.9+rke2r1",
```
Comment (MbolotSuse) [inline pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go:98]: I see we have examples for the rke2/k3s version strings - just to confirm, are the EKS/GKE/AKS version strings just the basic version or do they have a more complex syntax as well? What about generic imported clusters?

**Review by MbolotSuse — CHANGES_REQUESTED** [review]
Some small nits, but the overall approach looks good.

File: pkg/controllers/managementuser/rbac/podsecuritypolicy/serviceaccount.go (hunk 1 above)
Comment (maxsokolovsky) [reply pkg/controllers/managementuser/rbac/podsecuritypolicy/serviceaccount.go:126]: Agreed, will adjust!

File: pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+			wantErr: false,
+		},
+		{
+			version: "1.25.9",
+			wantErr: true,
+		},
+		{
+			version: "v1.25.9",
+			wantErr: true,
+		},
```
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go:65]: These aren't duplicated, one has a leading "v", the other doesn't.

File: pkg/controllers/managementuser/pspdelete/pspdelete.go [hunk 5]
```diff
@@ -25,6 +28,18 @@ type handler struct {
 
 func Register(ctx context.Context, userContext *config.UserContext) {
 	starter := userContext.DeferredStart(ctx, func(ctx context.Context) error {
+		clusterName := userContext.ClusterName
+		logrus.Infof("Checking cluster [%s] compatibility before registering pspdelete controller.", clusterName)
+		clusterLister := userContext.Management.Management.Clusters("").Controller().Lister()
+		err := podsecuritypolicy.CheckClusterVersion(clusterName, clusterLister)
+		if err != nil {
+			if errors.Is(err, podsecuritypolicy.ErrClusterVersionIncompatible) {
+				logrus.Errorf("%v - will not register pspdelete controller for cluster [%s].", err, clusterName)
```
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/controllers/managementuser/pspdelete/pspdelete.go:37]: Agreed, I'll switch it to Info.

File: pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+			} else if name == "not ready" {
+				return &v3.Cluster{Status: v3.ClusterStatus{}}, nil
+			} else {
+				return &v3.Cluster{Status: v3.ClusterStatus{Version: &version.Info{GitVersion: kubernetesVersion}}}, nil
+			}
+		},
+	}
+}
+
+func TestCheckClusterVersionFailsForVersionsThatCannotBeParsed(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (MbolotSuse) [reply pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go:28]: > The error kinds are different for all three tests.
Right, but that just means you need different output from the function, not that you are testing something fundamentally different.

> Uniting them into one test is very messy, the assertion logic would be hard to write and read. 
I don't agree, I think the tests would look very similar to what they look like today, only with an additional value to indicate the error type desired.

> It's perfectly fine to split tests into small chunks of quick, readable statements about expected behavior.

I'm not sure I agree here either. Consider this scenario - suppose that I update this function to add a new dependency. Now, instead of updating one test, I need to look through 3 different tests, each with a different structure and different mocked dependencies. 

In short, this structure makes it harder to update the tests and the related code in the future.

[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go:28]: Yes, but this is perfectly fine. The failures will also alert you to the need to adjust the tests. And it all encourages to have fewer dependencies and simpler test setup.

Consider these tests here: https://github.com/golang/go/blob/ab5bd15941f3cea3695338756d0b8be0ef2321fb/src/bufio/bufio_test.go#L289.
The logic of each test is small and separate. Alternatively, the authors could have put all setup into a `setup` func, a field of test struct, but it'd be a bit clunky.

File: pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go (hunk 3 above)
Comment (maxsokolovsky) [reply pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go:98]: I tried creating a managed DigitalOcean cluster and saw that its version on the DO portal appears to be `1.26.9-do.0`.
But when I import it into Rancher, I see it as `1.26.9` in Rancher. I will test other imported clusters.

File: pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+		},
+		// cloud provider version strings
+		{
+			version: "v1.27.3-gke.100",
+			wantErr: true,
+		},
+		{
+			version: "v1.26.9-eks-f8587cb",
+			wantErr: true,
+		},
```
Comment (MbolotSuse) [inline pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go:117]: I think it would also be good to test positive test cases as well (i.e. `v1.25.9-eks-f8587cb`)

File: pkg/controllers/managementuser/rbac/podsecuritypolicy/version.go [hunk 8]
```diff
@@ -4,25 +4,32 @@ import (
...
 
-var errVersionIncompatible = errors.New("podsecuritypolicies are not available in Kubernetes v1.25 and above")
-var clusterVersionCheckErrorString = "failed to check cluster version compatibility for podsecuritypolicy controllers: %v"
+var ErrClusterVersionIncompatible = errors.New("podsecuritypolicies are not available in Kubernetes v1.25 and above")
 
-// checkClusterVersion tries to fetch a cluster by name, extract its Kubernetes version,
-// and check if the version is less than v1.25.
-func checkClusterVersion(clusterName string, clusterLister v3.ClusterLister) error {
+// CheckClusterVersion tries to fetch a management.cattle.io Cluster by name, extract its Kubernetes version,
+// and check if the cluster supports PodSecurityPolicies.
```
Comment (MbolotSuse) [inline pkg/controllers/managementuser/rbac/podsecuritypolicy/version.go:14]: Nit: Would be good to indicate in the doc string that this returns a `ErrClusterVersionIncompatible` if we can identify the version, and confirm the cluster doesn't support PSPs, since that's a pretty critical part of the interface.

**Review by MbolotSuse — CHANGES_REQUESTED** [review]
2 small test cases that we should add, a nit on the docstring for the function, and there are threads on the testing struct that are not resolved.

**Review by MbolotSuse — APPROVED** [review]

File: pkg/controllers/managementuser/pspdelete/pspdelete.go [hunk 9]
```diff
@@ -25,6 +29,18 @@ type handler struct {
 
 func Register(ctx context.Context, userContext *config.UserContext) {
 	starter := userContext.DeferredStart(ctx, func(ctx context.Context) error {
+		clusterName := userContext.ClusterName
+		logrus.Infof("Checking cluster [%s] compatibility before registering pspdelete controller.", clusterName)
```
Comment (KevinJoiner) [inline pkg/controllers/managementuser/pspdelete/pspdelete.go:33]: We will always log the result of the version check, So we do not gain much from telling the user we are about to check and then giving them the result in the following log statement.
```suggestion
```

File: pkg/controllers/managementuser/pspdelete/pspdelete.go [hunk 10]
```diff
@@ -25,6 +29,18 @@ type handler struct {
...
+		clusterLister := userContext.Management.Management.Clusters("").Controller().Lister()
+		err := podsecuritypolicy.CheckClusterVersion(clusterName, clusterLister)
+		if err != nil {
+			if errors.Is(err, podsecuritypolicy.ErrClusterVersionIncompatible) {
+				logrus.Infof("%v - will not register pspdelete controller for cluster [%s].", err, clusterName)
```
Comment (KevinJoiner) [inline pkg/controllers/managementuser/pspdelete/pspdelete.go:38]: During normal execution, I don't think it is helpful for the user to know that the PSP controller was not registered, which makes me lean towards this being a debug message.
```suggestion
				logrus.Debugf("%v - will not register pspdelete controller for cluster [%s].", err, clusterName)
```
I don't want to spam comments but this applies to the other version checks as well.

File: pkg/controllers/managementuser/pspdelete/pspdelete.go [hunk 11]
```diff
@@ -25,6 +29,18 @@ type handler struct {
...
+				return nil
+			}
+			return fmt.Errorf("unable to parse version of cluster %s: %w", clusterName, err)
+		}
+		logrus.Infof("Cluster [%s] is compatible with PSPs, will run pspdelete controller.", clusterName)
```
Comment (KevinJoiner) [inline pkg/controllers/managementuser/pspdelete/pspdelete.go:43]: We do not usually log when controllers start, so this can be debug or even trace.
```suggestion
		logrus.Debugf("Cluster [%s] is compatible with PSPs, will run pspdelete controller.", clusterName)
```

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

**Comment by KevinJoiner** [comment]
@gregsidelinger / @maxsokolovsky 

> So this looks to address the two issues around PSPs when they don't exist in 1.25+ but any idea why this is breaking the controller reconcile loops for the other objects?

We start all of our controllers at the same time via [SharedControllerFactory.Start(...)](https://github.com/rancher/lasso/blob/d684fdeb6f29e221289c6f59fdc390b6adbaaccf/pkg/controller/sharedcontrollerfactory.go#L114C37-L114C37)

This function loops through all of the informers (caches) and attempts to populate them with a list call and then start a watch command to K8s, which sends the controller updates. This is all done via the [sharedCacheFactory.WaitForCacheSync(...)](https://github.com/rancher/lasso/blob/master/pkg/cache/sharedinformerfactory.go#L112) function.
While [looping through](https://github.com/rancher/lasso/blob/master/pkg/cache/sharedinformerfactory.go#L129C3-L129C3) and starting informers, we get blocked when we reach the PSP informer because the resource does not exist.

So depending on the order of this loop some of the other informers may have already started thus the reconcile logic works as expected and other are stuck waiting on the PSP informer.

File: pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go [hunk 12]
```diff
@@ -90,16 +107,37 @@ func TestCheckClusterVersion(t *testing.T) {
...
-			println(tt.version)
+			t.Parallel()
+			clusterLister := newClusterLister(tt.version)
 			err := checkClusterVersion("test", clusterLister)
 			if tt.wantErr {
-				println(err.Error())
-				assert.Error(t, err, "Expected checkClusterVersion to raise error.")
-				return
+				require.Error(t, err)
+				require.ErrorIs(t, err, errVersionIncompatible)
```
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go]: I also want to assert that the type of the error is one I expect and not any other.

File: pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go [hunk 13]
```diff
@@ -90,16 +107,37 @@ func TestCheckClusterVersion(t *testing.T) {
...
+			if !tt.wantErr {
+				require.NoError(t, err)
 			}
 		})
 	}
+}
+
+func TestCheckClusterVersionFailsWhenItCannotFetchVersion(t *testing.T) {
+	t.Parallel()
+	t.Run("version check fails when it can't get cluster", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go:142]: I think the main premise is common between these two - can't get the version. Subtests capture this fine, though I didn't even have to use those, could just have multiple assertions in a single test. But subtests add a bit of clarity.
But, sure, I can split them into separate tests...

**Review by MbolotSuse — APPROVED** [review]

**Review by KevinJoiner — APPROVED** [review]

File: pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (KevinJoiner) [reply pkg/controllers/managementuser/rbac/podsecuritypolicy/version_test.go]: Yeah, that was my point, but this is fine, line 116 is not needed since if the err is nil it will fail on 117.
